### ISSUE TITLE
GCPLog stops consuming messages

### ISSUE THREAD

**Issue by SvenW** [grafana/loki#6505]
**Describe the bug**
I have three instances of promtail running with the gcplog configuration and I noticed that they all had stopped consuming messages during the weekend. They all got `caller=target.go:111 msg="failed to receive pubsub messages" error="context deadline exceeded"` at different times during the day and after that stopped consuming messages completely. I'm running version `2.5.0`. 
I narrowed down the issue to this part of the code. If `sub.Receive` returns an error the gcplog target will stop consuming messages and cancel its context according to these lines 👇 

https://github.com/grafana/loki/blob/d9a522e5e6f93965324410d4b75a4cabd9ea1ac6/clients/pkg/promtail/targets/gcplog/target.go#L102-L115

**Expected behaviour**
- The gcplog target should retry on timeout issues 

**Environment**
* Infrastructure: GKE 1.22
* Promtail 2.5.0

**Comment by kavirajk** [comment]
@SvenW thanks for reporting this.

Can you tell more info about the status of the targets during that time.? You can access it via `/targets` page of promtail (default port 9080)

want to understand if its still active on target page, while it's not receiving messages.

**Comment by SvenW** [comment]
> @SvenW thanks for reporting this.
> 
> Can you tell more info about the status of the targets during that time.? You can access it via `/targets` page of promtail (default port 9080)
> 
> want to understand if its still active on target page, while it's not receiving messages.

@kavirajk as soon as I get one pod in the same state again I will check that. That was my thought as well. If the target isn't active any longer I would guess that promtail itself would do something about it

**Comment by kavirajk** [comment]
https://github.com/grafana/loki/blob/main/clients/pkg/promtail/targets/gcplog/target.go#L140-L145

ideally, gcplog should stop scraping because of any transient errors on Receive call. I'm curious if i't may be becoming in-active for any other reasons.

> @kavirajk as soon as I get one pod in the same state again I will check that.

Sure. thanks

**Comment by kavirajk** [comment]
btw. you can check the metric `gcplog_target_last_success_scrape` to see when was it last successfully scrapped and check the promtail logs during that time if anything else happened.

**Comment by SvenW** [comment]
> btw. you can check the metric `gcplog_target_last_success_scrape` to see when was it last successfully scrapped and check the promtail logs during that time if anything else happened.

The logs are completely empty after I get the timeout message. The metric I've just fixed so will verify that at the same time as I verify the `/targets` endpoint

**Comment by SvenW** [comment]
So tonight I had one more pod that got the same message. The targets page looks as the image below

![Screenshot 2022-06-28 at 08 26 14](https://user-images.githubusercontent.com/1511150/176108279-0a5174c3-d8ba-42bc-b26d-ac57d3306d02.png)

The metric `gcplog_target_parsing_errors_total` has been increased but the `gcplog_target_last_success_scrape` metric hasn't been set at all. Do you have anything else you want me to check @kavirajk?

**Comment by kavirajk** [comment]
Thanks for the more info @SvenW. So the target page shows Ready=True even after it stopped scraping correct?

Looks like a bug then. I think I will investigate it later today to see if I can find some fix.

**Comment by SvenW** [comment]
> Thanks for the more info @SvenW. So the target page shows Ready=True even after it stopped scraping correct?
Exactly, it is now showing Ready=True but doesn't scrape any longer

**Comment by SvenW** [comment]
@kavirajk just want to make sure that this issue isn't going to be marked as stale. Let me know if you need something else

**Comment by bizrad** [comment]
I'm also seeing this issue and error message. The failure appears to happen exactly every 30 days running promtail gcplog in our GKE k8s system. I wonder if there is some token expiration happening here?

**Comment by dmitri-lerko** [comment]
We are ingesting all of GCP Cloud Logging via promtail. At ~1MB/s it would stop receiving few times a day. I've investigated the issue and found out that Elastic Beats guys had a similar issue and it got solved upstream by PubSub library maintainers. I am now running a latest PubSub library for 5 days and had 0 issues. As an extra bonus, throughput per instance has increased by ~33%.

Happy to put together a PR to bump up PubSub.
